src/wp-includes/media.php

```diff
@@ -793,6 +803,356 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+		$img_ratio_compare = $img['height'] / $img['width'];
+
+		// If the new ratio differs by less than 0.01, use it.
+		if ( abs( $img_ratio - $img_ratio_compare ) < 0.01 ) {
+			$sources[ $img['width'] ] = path_join( dirname( $img_url ), $img['file'] ) . ' ' . $img['width'] .'w';
+		}
+	}
+
+	/**
+	 * Filter the output of wp_get_srcset_array().
+	 *
+	 * @since 4.4.0
+	 *
+	 * @param array        $sources  An array of image urls and widths.
+	 * @param int          $id       Attachment ID for image.
+	 * @param array|string $size     Size of image, either array or string.
+	 */
+	return apply_filters( 'wp_attachment_img_srcset', $sources, $id, $size );
+}
+
+/**
+ * Get the value for the 'srcset' attribute.
+ *
+ * @since 4.4.0
+ *
+ * @param int    $id   Image attachment ID.
+ * @param string $size Optional. Name of image size. Default value: 'medium'.
+ * @return string|bool A 'srcset' value string or false.
+ */
+function wp_attachment_img_srcset( $id, $size = 'medium' ) {
+	$srcset_array = wp_attachment_img_srcset_array( $id, $size );
+
+	// Only return a srcset value if there is more than one source.
+	if ( count( $srcset_array ) <= 1 ) {
+		return false;
+	}
+
+	return implode( ', ', $srcset_array );
+}
+
+/**
+ * Return a source size attribute for an image from an array of values.
+ *
+ * @since 4.4.0
+ *
+ * @param int    $id   Image attachment ID.
+ * @param string $size Optional. Name of image size. Default value: 'medium'.
+ * @param array  $args {
+ *     Optional. Arguments to retrieve posts.
+ *
+ *     @type array|string $sizes An array or string containing of size information.
+ *     @type int          $width A single width value used in the default `sizes` string.
+ * }
+ * @return string|bool A valid source size value for use in a 'sizes' attribute or false.
+ */
+function wp_attachment_img_sizes( $id, $size = 'medium', $args = null ) {
+
+	// Try to get the image width from `$args` before calling `image_downsize()`.
+	if ( is_array( $args ) && ! empty( $args['width'] ) ) {
+		$img_width = (int) $args['width'];
+	} elseif ( $img = image_downsize( $id, $size ) ) {
+		$img_width = $img[1];
+	}
+
+	// Bail early if `$image_width` isn't set.
+	if ( ! $img_width ) {
+		return false;
+	}
+
+	// Set the image width in pixels.
+	$img_width = $img_width . 'px';
+
+	// Set up our default values.
+	$defaults = array(
+		'sizes' => array(
+			array(
+				'size_value' => '100vw',
+				'mq_value'   => $img_width,
+				'mq_name'    => 'max-width'
+			),
+			array(
+				'size_value' => $img_width
+			),
+		)
+	);
+
+	$args = wp_parse_args( $args, $defaults );
+
+	/**
+	* Filter arguments used to create 'sizes' attribute.
+	*
+	* @since 2.4.0
+	*
+	* @param array   $args  An array of arguments used to create a 'sizes' attribute.
+	* @param int     $id    Post ID of the original image.
+	* @param string  $size  Name of the image size being used.
+	*/
+	$args = apply_filters( 'wp_image_sizes_args', $args, $id, $size );
```

Comment:
wp_resp_image_sizes_args?  Or what are the args really?  Maybe we can use that instead to avoid being generic?